contents/blog/clickhouse-materialized-columns.md

```diff
@@ -0,0 +1,134 @@
+
+- Took 3433ms
+- Read 79.17 GiB from disk
+
+To dig even deeper, we can use clickhouse-flamegraph to peek into what the CPU is doing during query execution.
+
+[![Flamegraph](../images/blog/clickhouse-materialized-columns/query-json-extract-CPU.svg)](../images/blog/clickhouse-materialized-columns/query-json-extract-CPU.svg)
+
+From this we can see that the ClickHouse server CPU is spending most of its time parsing JSON.
+
+The typical solution would be to extract $current_url to a separate column. This would get rid of the JSON parsing and reduce the amount of data read from disk.
+
+However, in this particular case it wouldn’t work because:
+
+1. The data is passed from users - meaning we’d end up with millions (!) of unique columns
+2. This would complicate live data ingestion a lot, introducing new and exciting race conditions
+
+
+## Enter materialized columns
+
+Turns out, that's exactly the problems materialized columns can help solve.
+
+```sql
+ALTER TABLE events
+ADD COLUMN mat_$current_url
+VARCHAR MATERIALIZED JSONExtractString(properties_json, '$current_url')
+```
+
+This will create a new column that will be automatically filled for incoming data, creating a new file on disk. The data is automatically filled during `INSERT` statements, so data ingestion does not need to change.
+
+The trade-off is more data being stored on disk. In practice ClickHouse compresses data very well, making this a worthwhile trade-off. On our test dataset, mat_$current_url is only 1.5% the size of `properties_json` on disk with a 10x compression ratio. Other properties which have lower cardinality can achieve even better compression (we’ve seen up to 100x)!
+
+Just creating the column is not enough, since for old data queries would still resort to using a `JSONExtract`. For this reason, you want to backfill data. The easiest way currently is to run [OPTIMIZE](https://clickhouse.tech/docs/en/sql-reference/statements/optimize/) command:
+
+```sql
+OPTIMIZE TABLE events FINAL
+```
+
+After backfilling, running the updated query speeds things up significantly:
+
+```sql
+SELECT count(*)
+FROM events
+WHERE event = '$pageview'
+  AND mat_$current_url = 'https://app.posthog.com/login'
+  AND timestamp >= '2021-08-01'
+  AND timestamp < '2021-09-01'
+```
+
+Looking at `system.query_log`, the new query:
+
+- took 980ms (**71%/3.4x improvement**)
+- read 14.36 GiB from disk (**81%/5x improvement improvement**)
+
+The wins are even more magnified if more than one property filter is used at a time.
+
+
+## Backfilling efficiently
+
+Using `OPTIMIZE TABLE` after adding columns is often not a good idea, since it will involve a lot of I/O as the whole table gets rewritten.
+
+As of writing, there's a a feature request on [Github](https://github.com/ClickHouse/ClickHouse/issues/27730) for adding specific commands for this.
+
+As a work-around you can:
+1. Temporarily set the column to use `DEFAULT` instead of `MATERIALIZED`
```

Comment:
Hi! Nice article. One thing I would point out is that at Altinity we often recommend that people just use DEFAULT from the start, as follows.  It works like MATERIALIZED in the sense that if the value isn't actually on disk ClickHouse will compute it. 

1.) Add the new column as follows. 
```
ALTER TABLE events
ADD COLUMN mat_$current_url
VARCHAR DEFAULT JSONExtractString(properties_json, '$current_url')
```
2. Now you can backfill as follows across the entire table. 
```
ALTER TABLE events UPDATE mat_$current_url WHERE 1=1
```